### DISCUSSION TITLE
What exactly does `defer` do?

### DISCUSSION THREAD

**Discussion by bnmoch3** [p-org/P#515] [Q&A] [answered]
What does defer do? I couldn't get much details in the docs. My assumption is that rather than drop/ignore an event, it queues it back with the requirement that it should only be popped off the queue when in any other state other than the current one

**Comment by ankushdesai** [comment, marked as answer]
Defer basically defers the dequeue of that event until it is non-deferred in a state. The position of the event does not change in the input buffer. Each dequeue event goes over the queue from the front and removes the first event that is not deferred, keeping rest of the queue unchanged.

**Reply by bnmoch3** [reply]
At what point does a deferred event become non-deferred, is it whenever the state changes?

**Reply by ankushdesai** [reply]
Yes, when you enter a state that does not defer that event.

**Reply by bnmoch3** [reply]
Cool, got it. I appreciate the help
